src/core/AutomationNode.cpp

```diff
@@ -0,0 +1,106 @@
+/*
+ * AutomationPattern.cpp - Implementation of class AutomationNode which
+ *                         holds information on a single automation pattern node
+ *
+ * Copyright (c) 2020 Ian Caio <iancaio_dev/at/hotmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "AutomationNode.h"
+
+#include "AutomationPattern.h"
+
+
+// Dummy constructor for the QMap
+AutomationNode::AutomationNode() :
+	m_pattern(nullptr),
+	m_key(0),
+	m_inValue(0),
+	m_outValue(0),
+	m_inTangent(0),
+	m_outTangent(0)
+{
+}
+
+AutomationNode::AutomationNode(AutomationPattern* pat, float value, int key) :
+	m_pattern(pat),
+	m_key(key),
+	m_inValue(value),
+	m_outValue(value),
+	m_inTangent(0),
+	m_outTangent(0)
+{
+}
+
+AutomationNode::AutomationNode(AutomationPattern* pat, float inValue, float outValue, int key) :
+	m_pattern(pat),
+	m_key(key),
+	m_inValue(inValue),
+	m_outValue(outValue),
+	m_inTangent(0),
+	m_outTangent(0)
+{
+}
+
+/**
+ * @brief Sets the inValue of an automation node
+ * @param Float value to be assigned
+*/
+void AutomationNode::setInValue(float value)
+{
+	m_inValue = value;
+
+	// Recalculate the tangents from neighbor nodes
+	AutomationPattern::timeMap & tm = m_pattern->getTimeMap();
+
+	// Get an iterator pointing to this node
+	AutomationPattern::timeMap::iterator it = tm.lowerBound(m_key);
+	// If it's not the first node, get the one immediately behind it
+	if (it != tm.begin())
+	{
+		--it;
+	}
+
+	// Generate tangents from the previously, current and next nodes
+	m_pattern->generateTangents(it, 3);
+}
+
+/**
+ * @brief Sets the outValue of an automation node
+ * @param Float value to be assigned
+*/
+void AutomationNode::setOutValue(float value)
+{
+	m_outValue = value;
+
+	// Recalculate the tangents from neighbor nodes
+	AutomationPattern::timeMap & tm = m_pattern->getTimeMap();
+
+	// Get an iterator pointing to this node
+	AutomationPattern::timeMap::iterator it = tm.lowerBound(m_key);
+	// If it's not the first node, get the one immediately behind it
```

Comment:
Does this logic really still work? Testing seems to imply it does, but I'm not sure how. When there's a jump in the node, shouldn't we treat outNodes as if they were the first node and inNodes as if they were the last node? Intuitively (to me) the node shouldn't "see" nodes on the other side of the jump.

Reply:
It does because `generateTangents` was changed, so when calculating the tangents of a node it considers the right side and left side of the pattern separately. If `inValue` and `outValue` on the node that is having its tangents calculated are the same, then it considers the curve continuous from the previous node's `outValue` up to the next node's `inValue`. If the `inValue` and `outValue` from the node being calculated are different, than two separate curves are considered: One on the left side and one on the right side. Because of that you'll notice that when dragging an `outValue`, the moment it passes the level of `inValue` the curve changes a little suddenly (because when `outValue` matched `inValue` the curve was considered continuous, so the tangents were different).

It's a little hard explain writing, but if it sounded confusing I can try to elaborate better!